log/DESIGN.md

```diff
@@ -0,0 +1,446 @@
+
+[`Context` associated with the `LogRecord`](https://opentelemetry.io/docs/specs/otel/context/)
+is accepted as a `context.Context` method argument.
+
+Calls to `Emit` are supposed to be on the hot path.
+Therefore, in order to reduce the number of heap allocations,
+the [`LogRecord` abstraction](https://opentelemetry.io/docs/specs/otel/logs/bridge-api/#emit-a-logrecord),
+is defined as a `struct`:
+
+```go
+type Record struct {
+	Timestamp         time.Time
+	ObservedTimestamp time.Time
+	Severity          Severity
+	SeverityText      string
+	Body              string
+	Attributes        []attribute.KeyValue
+}
+```
+
+[`Timestamp`](https://opentelemetry.io/docs/specs/otel/logs/data-model/#field-timestamp)
+is defined as `time.Time` type.
+
+[`ObservedTimestamp`](https://opentelemetry.io/docs/specs/otel/logs/data-model/#field-observedtimestamp)
+is defined as `time.Time` type.
+
+[`SeverityNumber`](https://opentelemetry.io/docs/specs/otel/logs/data-model/#field-severitynumber)
+is defined as a type and constants:
+
+```go
+type Severity int
+
+const (
+	SeverityTrace Severity = iota + 1
+	SeverityTrace2
+	SeverityTrace3
+	SeverityTrace4
+	SeverityDebug
+	SeverityDebug2
+	SeverityDebug3
+	SeverityDebug4
+	SeverityInfo
+	SeverityInfo2
+	SeverityInfo3
+	SeverityInfo4
+	SeverityWarn
+	SeverityWarn2
+	SeverityWarn3
+	SeverityWarn4
+	SeverityError
+	SeverityError2
+	SeverityError3
+	SeverityError4
+	SeverityFatal
+	SeverityFatal2
+	SeverityFatal3
+	SeverityFatal4
+)
+```
+
+[`SeverityText`](https://opentelemetry.io/docs/specs/otel/logs/data-model/#field-severitytext)
+is defined as `string` type.
+
+[`Body`](https://opentelemetry.io/docs/specs/otel/logs/data-model/#field-body)
+is defined as `string` type as the specification says:
+
+> First-party Applications SHOULD use a string message.
+
+[Log record attributes](https://opentelemetry.io/docs/specs/otel/logs/data-model/#field-attributes)
+are defined a regular slice of `attribute.KeyValue`.
+The users can use [`sync.Pool`](https://pkg.go.dev/sync#Pool)
+for reducing the number of allocations when passing attributes.
+
+Implementation requirements:
+
+- The [specification requires](https://opentelemetry.io/docs/specs/otel/logs/bridge-api/#concurrency-requirements)
+  the method to be safe to be called concurrently.
+
+- The method should not interrupt the record processing if the context is canceled
+  per ["ignoring context cancellation" guideline](../CONTRIBUTING.md#ignoring-context-cancellation).
+
+- The method handle the trace context passed via `ctx` argument in order to meet the
+  [specification's SDK requirement](https://opentelemetry.io/docs/specs/otel/logs/sdk/#readablelogrecord)
+  to populate the trace context fields from the resolved context.
+
+- The method should not modify the record's attribute as it may be reused by caller.
+
+- The method should copy the record's attributes in case of asynchronous processing
```

Comment:
I wonder if this leads overall to better performance compared to just letting GC do its job. Has this been benchmarked on large and complex nested attribute values?

Reply:
It is not about performance. It is about correctness in  order to avoid data races.

Reply:
This is true but it is a general guideline for API implementations (e.g. for resting). It is possible that the SDK would always create need attributes. Probably the SDK will use a sync.Pool to decrease numer of allocs and decrease the presure on GC. It will be explored when working on Logs SDK design+prototype.

Reply:
@tigrannajaryan Resolving. Feel free to unresolve if you do not agree with my arguments.

Reply:
> guarantee that the attributes won't change after the call returns

I do not think it is possible.

Reply:
SIG meeting:

We agreed that it would be good to ask users to not mutate the record after passing it to `Emit` similarly to https://pkg.go.dev/log/slog#GroupValue. We can still keep the recommendation API to copy the copy as some users may not read the docs.

Reply:
Addressed 60cb10a3

Thank you